tests/istanbul/eip152_blake2/test_blake2.py

```diff
@@ -0,0 +1,547 @@
+"""
+abstract: Tests [EIP-152: BLAKE2 compression precompile](https://eips.ethereum.org/EIPS/eip-152)
+    Test cases for [EIP-152: BLAKE2 compression precompile](https://eips.ethereum.org/EIPS/eip-152).
+"""
+
+from dataclasses import dataclass
+
+import pytest
+
+from ethereum_test_tools import (
+    Account,
+    Alloc,
+    Environment,
+    StateTestFiller,
+    TestParameterGroup,
+    Transaction,
+)
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+REFERENCE_SPEC_GIT_PATH = "EIPS/eip-152.md"
+REFERENCE_SPEC_VERSION = "5510973b40973b6aa774f04c9caba823c8ff8460"
+
+
+@dataclass(kw_only=True, frozen=True, repr=False)
+class Blake2Input(TestParameterGroup):
+    """
+    Helper class that defines the BLAKE2 precompile inputs and creates the
+    call data from them. Returns all inputs encoded as bytes.
+
+    Attributes:
+        rounds_length (int): An optional integer representing the bytes length
+            for the number of rounds. Defaults to the expected length of 4.
+        rounds (int): An integer representing the number of rounds.
+        h (str): A hex string that represents the state vector.
+        m (str): A hex string that represents the message block vector.
+        t_0 (str): A hex string that represents the first offset counter.
+        t_1 (str): A hex string that represents the second offset counter.
+        f (bool): A boolean that represents the final block indicator flag.
+
+    """
+
+    rounds_length: int = 4
+    rounds: int
+    h: str
+    m: str
+    t_0: str
+    t_1: str
+    f: bool = True
```

Comment:
```suggestion
    rounds: int | str
    h: Tuple[int, ...] | str
    m: Tuple[int, ...] | str
    t_0: int | str
    t_1: int | str
    f: bool | str = True
```
We could change the types here to either int or string:
- int: the value gets converted to bytes using the endianness specified in the spec automatically.
- str: the string is parsed as hex bytes and copied verbatim into the formatted precompile call data that is generated in `create_blake2_tx_data`.

And the reasoning is that it's easier to differentiate between the cases that call the precompile with the correct data size from those that don't IMO. 

Maybe for `h` and `m` it makes less sense because the precompile expects a tuple of ints, but for `rounds`, `t_0`, `t_1` and `f` I think it should be ok and it should be a bit easier to read.

Reply:
I like that idea so the test parameters are all consistent integers. To be certain, the parameterized tests will end up with h, m, t_* integer values instead? For example:
```
            Blake2bInput(
                rounds=0,
                h=[
                    32922972323324069763321526652467788264663358122255454854356121218204912930725,
                    94764667551462817271282565646542853509102241772507082255866415388681393004635,
                ],
                m=[6382179, 0, 0, 0],
                t_0=3,
                t_1=0,
            ),
```

Reply:
> I like that idea so the test parameters are all consistent integers. To be certain, the parameterized tests will end up with h, m, t_* integer values instead? I'm looking at how to encode the Tuple of integers together as one bytes value, is there a standard way of doing that?
> 
> For example:
> 
> ```
>             Blake2bInput(
>                 rounds=0,
>                 h=[
>                     32922972323324069763321526652467788264663358122255454854356121218204912930725,
>                     94764667551462817271282565646542853509102241772507082255866415388681393004635,
>                 ],
>                 m=[6382179, 0, 0, 0],
>                 t_0=3,
>                 t_1=0,
>             ),
> ```

Yeah, I think it makes more sense for `round` and `t_*` because they are smaller numbers, and to convert them is a single `to_bytes`. Following the [EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md#specification) it should be something like:
```python
self.rounds.to_bytes(length=4, byteorder="big")
self.t_0.to_bytes(length=8, byteorder="little")
self.t_1.to_bytes(length=8, byteorder="little")
```
> I'm looking at how to encode the Tuple of integers together as one bytes value, is there a standard way of doing that?

I was thinking about using `join` and `to_bytes` for each integer in the Tuple, and use the length and byte order in the EIP.

So e.g. for `h`, "8 unsigned 64-bit little-endian words" in the EIP, the integer representation could be `(0, 0, 0, 0, 0, 0, 0, 0)`, so eight numbers where the max is `2**64-1`, and then I think maybe using
```python
b"".join(i.to_bytes(length=8, byteorder="little") for i in self.h)
```

*But*... I think the main annoyance right now is taking the strings we have and converting that to integer tuples, so I'm not sure if it's worth it since `m` and `h` is not something that can be easily readable as `rounds` is for example.